### PR TITLE
IOSXE/show ip bgp vpnv4 vrf user summary

### PR DESCRIPTION
**Pull request by qduk**
## Description
Parser created for show ip bgp vpnv4 vrf user summary

## Motivation and Context
Takes the unstructured data from the output and returns structured data

## Screenshots:
Capture

## Checklist:
- [x] I have updated the changelog.
- [x] All new and existing tests passed.
- [x] All new code passed compilation.

### CODE REVIEW

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 1]
```diff
@@ -110,5 +110,252 @@ def cli(self, output = None):
...
+
         return super().cli(output = show_output)
+
+
+# =======================================
+# Schema for:
+#  * 'show ip bgp vpnv4 vrf user summary'
+# =======================================
+class ShowIpBgpVpnv4VrfUserSummarySchema(MetaParser):
+    """Schema for show ip bgp vpnv4 vrf user summary."""
```
Comment (LenzDong) [inline src/genie/libs/parser/iosxe/show_ip.py:122]: Please pass ```user``` as a vrf argument
```
show ip bgp vpnv4 vrf {vrf} summary

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 2]
```diff
@@ -110,5 +110,252 @@ def cli(self, output = None):
...
+# =======================================
+# Parser for:
+#  * 'show ip bgp vpnv4 vrf user summary'
+# =======================================
+class ShowIpBgpVpnv4VrfUserSummary(ShowIpBgpVpnv4VrfUserSummarySchema):
+    """Parser for show ip bgp vpnv4 vrf user summary"""
+
+    cli_command = 'show ip bgp vpnv4 vrf user summary'
+
+    def cli(self, output=None):
```
Comment (LenzDong) [inline src/genie/libs/parser/iosxe/show_ip.py:195]: Please pass vrf name as an argument 
```
def cli(self, vrf, output=None):

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 3]
```diff
@@ -110,5 +110,252 @@ def cli(self, output = None):
...
+        # BGP using 38496 total bytes of memory
+        # BGP activity 854/668 prefixes, 2572/2305 paths, scan interval 60 secs
+        #
+        # Neighbor        V           AS MsgRcvd MsgSent   TblVer  InQ OutQ Up/Down  State/PfxRcd
+        # 10.199.228.201  4        65001  202074  202468     4571    0    0 18w1d          10
+        # 10.199.228.209  4        65001  499437  638231     4571    0    0 5w2d           58
+        # 10.199.228.217  4        65001  496306  634769     4571    0    0 5w2d           10
+
+        # BGP router identifier 10.19.228.213, local AS number 64106
+        p_bgp_router = re.compile(r"^BGP\s+router\s+identifier\s+(?P<rid>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}),\s+local\s+AS\s+number\s+(?P<as>\d+)$")
```
Comment (LenzDong) [inline src/genie/libs/parser/iosxe/show_ip.py:220]: rid can be simplify as ```(?P<rid>\S+)```

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 4]
```diff
@@ -110,5 +110,252 @@ def cli(self, output = None):
...
+        p_bgp_mem = re.compile(r"^BGP\s+using\s+(?P<mem>\d+)\s+total\s+bytes\s+of\s+memory$")
+
+        # BGP activity 854/668 prefixes, 2572/2305 paths, scan interval 60 secs
+        p_bgp_activity = re.compile(r"^BGP\s+activity\s+(?P<alloc_1>\d+)\/(?P<release_1>\d+)\s+prefixes,\s+(?P<alloc_2>\d+)\/(?P<release_2>\d+)\s+paths,\s+scan\s+interval\s+(?P<si>\d+)\s+secs$")
+
+        # 18 networks peaked at 07:08:57 Sep 16 2020 UTC (5w2d ago)
+        p_bgp_peaked = re.compile(r"^(?P<networks>\d+)\s+networks\s+peaked\s+at\s+(?P<date>[^(]+)\((?P<ago>\S+)\s+ago\)$")
+
+        # 10.199.228.201  4        65001  202074  202468     4571    0    0 18w1d          10
+        p_bgp_neighbor = re.compile(r"^(?P<neighbor>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\s+(?P<v>\d+)\s+(?P<as>\d+)\s+(?P<msgr>\d+)\s+(?P<msgs>\d+)\s+(?P<tblv>\d+)\s+(?P<inq>\d+)\s+(?P<outq>\d+)\s+(?P<ud>\S+)\s+(?P<state>\d+)$")
```
Comment (LenzDong) [inline src/genie/libs/parser/iosxe/show_ip.py:259]: neighbor can be simplify as ```(?P<neighbor>\S+)```

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 5]
```diff
@@ -110,5 +110,252 @@ def cli(self, output = None):
...
+                # 18 networks peaked at 07:08:57 Sep 16 2020 UTC (5w2d ago)
+                m_bgp_peaked = p_bgp_peaked.match(line)
+                bgp_dict.update({ "networks_peaked" : m_bgp_peaked.group("networks") })
+                bgp_dict.update({ "networks_peaked_time" : m_bgp_peaked.group("date").strip() })
+                bgp_dict.update({ "time_since_network_peak" : m_bgp_peaked.group("ago") })
+                continue
+            elif p_bgp_neighbor.match(line):
+                # 10.199.228.201  4        65001  202074  202468     4571    0    0 18w1d          10
+                m_bgp_neighbor = p_bgp_neighbor.match(line)
+                if not bgp_dict.get("neighbor_ips"):
```
Comment (LenzDong) [inline src/genie/libs/parser/iosxe/show_ip.py:347]: You can use ```setdefault```
```
neighbor_dict = bgp_dict.setdefault("neighbor_ips", {}).setdefault(m_bgp_neighbor.group("neighbor"), {})
neighbor_dict.update({ "version": int(m_bgp_neighbor.group("v")) })

File: src/genie/libs/parser/iosxe/show_ip.py (hunk 1 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_ip.py]: Updated

File: src/genie/libs/parser/iosxe/show_ip.py (hunk 2 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_ip.py]: Updated

File: src/genie/libs/parser/iosxe/show_ip.py (hunk 3 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_ip.py]: Updated

File: src/genie/libs/parser/iosxe/show_ip.py (hunk 4 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_ip.py]: Updated

File: src/genie/libs/parser/iosxe/show_ip.py (hunk 5 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_ip.py]: Updated

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 6]
```diff
@@ -110,5 +110,250 @@ def cli(self, output = None):
...
+# =======================================
+# Parser for:
+#  * 'show ip bgp vpnv4 vrf {vrf} summary'
+# =======================================
+class ShowIpBgpVpnv4VrfUserSummary(ShowIpBgpVpnv4VrfUserSummarySchema):
+    """Parser for show ip bgp vpnv4 vrf {vrf} summary"""
+
+    cli_command = 'show ip bgp vpnv4 vrf {vrf} summary'
+
+    def cli(self, vrf="", output=None):
```
Comment (LenzDong) [inline src/genie/libs/parser/iosxe/show_ip.py:195]: ```vrf``` should be a mandatory argument, please remove ```=""```:
```
def cli(self, vrf, output=None):

**Review by LenzDong — APPROVED** [review]

File: src/genie/libs/parser/iosxe/show_ip.py (hunk 6 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_ip.py:195]: Updated!

File: src/genie/libs/parser/iosxe/show_ip.py [hunk 7]
```diff
@@ -110,5 +110,250 @@ def cli(self, output = None):
...
+
         return super().cli(output = show_output)
+
+
+# =======================================
+# Schema for:
+#  * 'show ip bgp vpnv4 vrf {vrf} summary'
+# =======================================
+class ShowIpBgpVpnv4VrfUserSummarySchema(MetaParser):
+    """Schema for show ip bgp vpnv4 vrf {vrf} summary."""
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/show_ip.py:122]: this parser already exists... Please check if this is what you need and if yes, please decline this PR.
image

example:
```
>>> dev.parse('show ip bgp vpnv4 vrf VRF1 summary')

2020-10-30 11:06:34,313: %UNICON-INFO: +++ R1_xe: executing command 'show ip bgp vpnv4 vrf VRF1 summary' +++
show ip bgp vpnv4 vrf VRF1 summary
BGP router identifier 10.1.1.1, local AS number 65000
BGP table version is 26, main routing table version 26
2 network entries using 512 bytes of memory
2 path entries using 272 bytes of memory
2/2 BGP path/bestpath attribute entries using 608 bytes of memory
2 BGP extended community entries using 500 bytes of memory
0 BGP route-map cache entries using 0 bytes of memory
0 BGP filter-list cache entries using 0 bytes of memory
BGP using 1892 total bytes of memory
BGP activity 8/0 prefixes, 13/5 paths, scan interval 60 secs
2 networks peaked at 18:38:40 Oct 27 2020 UTC (2d20h ago)

Neighbor        V           AS MsgRcvd MsgSent   TblVer  InQ OutQ Up/Down  State/PfxRcd
10.2.2.2        4        65000       0       0        1    0    0 never    Idle
10.3.3.3        4        65000    4106    4516       26    0    0 2d20h           1

R1_xe#
{'bgp_id': 65000, 'vrf': {'VRF1': {'neighbor': {'10.2.2.2': {'address_family': {'vpnv4': {'version': 4, 'as': 65000, 'msg_rcvd': 0, 'msg_sent': 0, 'tbl_ver': 1, 'input_queue': 0, 'output_queue': 0, 'up_down': 'never', 'state_pfxrcd': 'Idle', 'route_identifier': '10.1.1.1', 'local_as': 65000, 'bgp_table_version': 26, 'routing_table_version': 26, 'attribute_entries': '2/2', 'prefixes': {'total_entries': 2, 'memory_usage': 512}, 'path': {'total_entries': 2, 'memory_usage': 272}, 'total_memory': 1892, 'activity_prefixes': '8/0', 'activity_paths': '13/5', 'scan_interval': 60, 'cache_entries': {'route-map': {'total_entries': 0, 'memory_usage': 0}, 'filter-list': {'total_entries': 0, 'memory_usage': 0}}, 'community_entries': {'total_entries': 2, 'memory_usage': 500}}}}, '10.3.3.3': {'address_family': {'vpnv4': {'version': 4, 'as': 65000, 'msg_rcvd': 4106, 'msg_sent': 4516, 'tbl_ver': 26, 'input_queue': 0, 'output_queue': 0, 'up_down': '2d20h', 'state_pfxrcd': '1', 'route_identifier': '10.1.1.1', 'local_as': 65000, 'bgp_table_version': 26, 'routing_table_version': 26, 'attribute_entries': '2/2', 'prefixes': {'total_entries': 2, 'memory_usage': 512}, 'path': {'total_entries': 2, 'memory_usage': 272}, 'total_memory': 1892, 'activity_prefixes': '8/0', 'activity_paths': '13/5', 'scan_interval': 60, 'cache_entries': {'route-map': {'total_entries': 0, 'memory_usage': 0}, 'filter-list': {'total_entries': 0, 'memory_usage': 0}}, 'community_entries': {'total_entries': 2, 'memory_usage': 500}}}}}}}}
>>>

**Review by tahigash — CHANGES_REQUESTED** [review]
it seems same parser already exists.

**Comment by qduk** [comment]
Parser already exists Closing PR.

**Pull request closed without merging**